File: 2023/day05/php/solution.php
<?php
/**
 * Advent of Code 2023 - Day 5: If You Give A Seed A Fertilizer
 */

/**
 * Parse input into seeds and list of maps.
 */
function parseInput(string $text): array {
    $sections = explode("\n\n", trim($text));

    // Parse seeds
    $seedPart = explode(': ', $sections[0])[1];
    $seeds = array_map('intval', preg_split('/\s+/', trim($seedPart)));

    // Parse maps
    $maps = [];
    for ($i = 1; $i < count($sections); $i++) {
        $lines = explode("\n", trim($sections[$i]));
        $ranges = [];
        // Skip header line
        for ($j = 1; $j < count($lines); $j++) {
            $parts = array_map('intval', preg_split('/\s+/', trim($lines[$j])));
            $ranges[] = [
                'dst_start' => $parts[0],
                'src_start' => $parts[1],
                'length' => $parts[2]
            ];
        }
        $maps[] = $ranges;
    }

    return [$seeds, $maps];
}

/**
 * Apply a single map to transform a value.
 */
function applyMap(int $value, array $ranges): int {
    foreach ($ranges as $range) {
        $srcStart = $range['src_start'];
        $srcEnd = $srcStart + $range['length'];
        if ($value >= $srcStart && $value < $srcEnd) {
            return $range['dst_start'] + ($value - $srcStart);
        }
    }
    return $value;
}

/**
 * Convert a seed number to a location number through all maps.
 */
function seedToLocation(int $seed, array $maps): int {
    $value = $seed;
    foreach ($maps as $mapRanges) {
        $value = applyMap($value, $mapRanges);
    }
    return $value;
}

/**
 * Part 1: Find the lowest location number for any initial seed.
 */
function part1(array $seeds, array $maps): int {
    $minLocation = PHP_INT_MAX;
    foreach ($seeds as $seed) {
        $location = seedToLocation($seed, $maps);
        if ($location < $minLocation) {
            $minLocation = $location;
        }
    }
    return $minLocation;
}

/**
 * Apply a map to a list of ranges, returning new ranges.
 */
function applyMapToRanges(array $inputRanges, array $mapRanges): array {
    $result = [];

    foreach ($inputRanges as $inputRange) {
        $start = $inputRange[0];
        $end = $inputRange[1];
        $remaining = [[$start, $end]];

        foreach ($mapRanges as $range) {
            $dstStart = $range['dst_start'];
            $srcStart = $range['src_start'];
            $length = $range['length'];
            $srcEnd = $srcStart + $length;

            $newRemaining = [];

            foreach ($remaining as $r) {
                $rStart = $r[0];
                $rEnd = $r[1];

                // Part before the map range (unmapped)
                if ($rStart < $srcStart) {
                    $newRemaining[] = [$rStart, min($rEnd, $srcStart)];
                }

                // Part within the map range (mapped)
                $overlapStart = max($rStart, $srcStart);
                $overlapEnd = min($rEnd, $srcEnd);
                if ($overlapStart < $overlapEnd) {
                    $offset = $dstStart - $srcStart;
                    $result[] = [$overlapStart + $offset, $overlapEnd + $offset];
                }

                // Part after the map range (unmapped)
                if ($rEnd > $srcEnd) {
                    $newRemaining[] = [max($rStart, $srcEnd), $rEnd];
                }
            }

            $remaining = $newRemaining;
        }

        // Any remaining parts are unmapped (identity)
        foreach ($remaining as $r) {
            $result[] = $r;
        }
    }

    return $result;
}

/**
 * Part 2: Find the lowest location for seed ranges.
 */
function part2(array $seeds, array $maps): int {
    // Convert seeds to ranges: pairs of [start, start + length]
    $ranges = [];
    for ($i = 0; $i < count($seeds); $i += 2) {
        $start = $seeds[$i];
        $length = $seeds[$i + 1];
        $ranges[] = [$start, $start + $length];
    }

    // Apply each map to the ranges
    foreach ($maps as $mapRanges) {
        $ranges = applyMapToRanges($ranges, $mapRanges);
    }

    // Find minimum start of any range
    $minStart = PHP_INT_MAX;
    foreach ($ranges as $range) {
        if ($range[0] < $minStart) {
            $minStart = $range[0];
        }
    }

    return $minStart;
}

function main(): void {
    $inputPath = __DIR__ . '/../input.txt';
    $text = file_get_contents($inputPath);

    [$seeds, $maps] = parseInput($text);

    echo 'Part 1: ' . part1($seeds, $maps) . "\n";
    echo 'Part 2: ' . part2($seeds, $maps) . "\n";
}

main();

File: 2023/day05/clojure/solution.clj
(ns solution
  (:require [clojure.string :as str]))

(def input (slurp "../input.txt"))

(defn parse-long-safe [s]
  "Parse a string to long, returning nil for empty strings"
  (when-not (str/blank? s)
    (parse-long s)))

(defn parse-seeds [line]
  "Parse the seeds line into a list of seed numbers"
  (->> (str/split (second (str/split line #": ")) #"\s+")
       (map parse-long-safe)
       (remove nil?)))

(defn parse-map-line [line]
  "Parse a map line into [dst-start src-start length]"
  (->> (str/split (str/trim line) #"\s+")
       (map parse-long)
       vec))

(defn parse-map-section [section]
  "Parse a map section into a vector of ranges"
  (let [lines (str/split-lines section)]
    (->> (rest lines)  ; Skip header line
         (map parse-map-line)
         vec)))

(defn parse-input [text]
  "Parse input into [seeds maps]"
  (let [sections (str/split (str/trim text) #"\n\n")
        seeds (parse-seeds (first sections))
        maps (mapv parse-map-section (rest sections))]
    [seeds maps]))

(defn apply-map-value [value ranges]
  "Apply a single map to transform a value"
  (if-let [match (first (filter (fn [[dst-start src-start length]]
                                   (and (>= value src-start)
                                        (< value (+ src-start length))))
                                 ranges))]
    (let [[dst-start src-start _] match]
      (+ dst-start (- value src-start)))
    value))

(defn seed-to-location [seed maps]
  "Convert a seed number to a location number through all maps"
  (reduce apply-map-value seed maps))

(defn part1 [seeds maps]
  "Find the lowest location number for any initial seed"
  (apply min (map #(seed-to-location % maps) seeds)))

(defn apply-map-to-ranges [input-ranges map-ranges]
  "Apply a map to a list of ranges, returning new ranges"
  (loop [remaining-inputs input-ranges
         mapped []
         unmapped []]
    (if (empty? remaining-inputs)
      ;; After all input ranges processed, remaining unmapped go through as identity
      (concat mapped unmapped)
      (let [[r-start r-end] (first remaining-inputs)
            rest-inputs (rest remaining-inputs)]
        ;; For this input range, try each map range
        (let [[new-mapped new-unmapped]
              (loop [remaining-maps map-ranges
                     current-remaining [[r-start r-end]]
                     result-mapped []]
                (if (or (empty? remaining-maps) (empty? current-remaining))
                  [result-mapped current-remaining]
                  (let [[dst-start src-start length] (first remaining-maps)
                        src-end (+ src-start length)
                        rest-maps (rest remaining-maps)
                        ;; Process all current remaining through this map range
                        processed
                        (for [[rs re] current-remaining]
                          (let [;; Part before map range (stays unmapped for now)
                                before (when (< rs src-start)
                                         [rs (min re src-start)])
                                ;; Overlap with map range (gets mapped)
                                overlap-start (max rs src-start)
                                overlap-end (min re src-end)
                                overlap (when (< overlap-start overlap-end)
                                          (let [offset (- dst-start src-start)]
                                            [(+ overlap-start offset)
                                             (+ overlap-end offset)]))
                                ;; Part after map range (stays unmapped for now)
                                after (when (> re src-end)
                                        [(max rs src-end) re])]
                            {:before before :overlap overlap :after after}))
                        ;; Collect mapped overlaps
                        new-results (->> processed
                                         (map :overlap)
                                         (remove nil?))
                        ;; Collect remaining unmapped (before + after)
                        new-remaining (->> processed
                                           (mapcat (fn [p] [(:before p) (:after p)]))
                                           (remove nil?))]
                    (recur rest-maps
                           new-remaining
                           (concat result-mapped new-results)))))]
          (recur rest-inputs
                 (concat mapped new-mapped)
                 (concat unmapped new-unmapped)))))))

(defn part2 [seeds maps]
  "Find the lowest location for seed ranges"
  ;; Convert seeds to ranges: pairs of [start, start + length)
  (let [seed-ranges (for [pair (partition 2 seeds)
                          :let [[start length] pair]]
                      [start (+ start length)])
        ;; Apply each map to the ranges
        final-ranges (reduce apply-map-to-ranges seed-ranges maps)]
    ;; Find minimum start of any range
    (apply min (map first final-ranges))))

;; Main execution
(let [[seeds maps] (parse-input input)]
  (println "Part 1:" (part1 seeds maps))
  (println "Part 2:" (part2 seeds maps)))
